File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        '''
        use three nested loop. 
        '''
        nums.sort()
        ans = []
        for i in range(len(nums)-2):
            if i>0 and nums[i] == nums[i-1]:
                continue
            l,r = i+1,len(nums)-1
            while l<r:
                if nums[l]+nums[i]+nums[r] > 0:
                    r-=1
                elif nums[l]+nums[i]+nums[r]<0:
                    l+=1
                else:
                    ans.append([nums[i],nums[l],nums[r]])
                    while r>l and nums[r] == nums[r-1]:
                        r-=1
                    r -= 1
        return ans
                
        

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a,b) => a-b)
    const ans = []
    for (let i=0;i<nums.length-2;i++){
        if (i>0){
            if(nums[i] == nums[i-1]){
                continue
            }
        }
        let curr = nums[i];
        let l = i+1;
        let r = nums.length - 1
        while (l<r){
            let val = nums[l] + nums[r] + curr
            if (val === 0){
                    ans.push([curr,nums[l],nums[r]]);
                const t = nums[r];
                while (r>l && t === nums[r]){
                    r -= 1
                }
            }
            else if(val<0){
                l += 1;
            }
            else{
                r -= 1;
            }
        }
    }
    return ans
    
};